### PR TITLE
Cube shop listing

### PR DESCRIPTION
**Pull request by carkod**
## Done

- Replace `cube.yaml` content with data from UA contracts API.
- Populate modules list with `/cube/microcerts.json`
- No actions yet, it'll make the PR too big and there are other issues opened for that.

## QA

- Check https://ubuntu-com-10421.demos.haus/cube/microcerts to see a new table with content
- Navigate to /cube/microcerts as a guest user, and make sure it redirects to USSO

## Issue / Card

Fixes https://github.com/canonical-web-and-design/commercial-squad/issues/256

## Screenshots

advantage-data

### CODE REVIEW

File: webapp/app.py [hunk 1]
```diff
@@ -793,6 +794,9 @@ def takeovers_index():
 app.add_url_rule("/cube", view_func=cube_home)
 app.add_url_rule("/cube/microcerts", view_func=cube_microcerts)
 app.add_url_rule("/cube/study/labs", view_func=cube_study_labs_button)
+app.add_url_rule(
+    "/cube/get-microcerts", view_func=get_microcerts, methods=["GET"]
```
Comment (anthonydillon) [inline webapp/app.py:798]: Shouldn't endpoint that return json end with a .json extension? And get is implied, right? What about: `/cube/microcerts.json`?

File: webapp/cube/views.py [hunk 2]
```diff
@@ -4,21 +4,30 @@
...
 
 from webapp.cube.api import BadgrAPI, EdxAPI
 from webapp.decorators import login_required
 from webapp.login import user_info
-
+from webapp.advantage.ua_contracts.api import UAContractsAPI
 
 CUBE_CONTENT = yaml.load(
-    Path("webapp/cube/content/cube.yaml").read_text(), Loader=yaml.Loader
+    Path("webapp/cube/content/cube-qa.yaml").read_text(), Loader=yaml.Loader
```
Comment (jpmartinspt) [inline webapp/cube/views.py:20]: Need to revert this to `cube.yaml`.

File: webapp/cube/views.py [hunk 3]
```diff
@@ -45,6 +54,8 @@
     edx_session,
 )
 
+ua_contracts_session = Session()
```
Comment (jpmartinspt) [inline webapp/cube/views.py:57]: add `talisker.requests.configure(ua_contracts_session)`

File: webapp/cube/views.py [hunk 4]
```diff
@@ -133,6 +144,149 @@ def cube_microcerts():
...
+    # backend_true requires query string,
+    # therefore we need context
+    # also this avoids loading contracts api where not required
+    is_test_backend = (
+        current_app.config["CONTRACTS_TEST_API_URL"]
+        if flask.request.args.get("test_backend", "false")
+        else current_app.config["CONTRACTS_LIVE_API_URL"]
+    )
+    sso_user = user_info(flask.session)
+    if not sso_user:
```
Comment (jpmartinspt) [inline webapp/cube/views.py:159]: Since this view is under a @login_required decorator it should be guaranteed that you have an `sso_user`

Comment (jpmartinspt) [reply webapp/cube/views.py:159]: Also kinda linked to this, if this is a JSON view, it should not be redirecting or returning HTML

File: webapp/cube/views.py [hunk 5]
```diff
@@ -133,6 +144,149 @@ def cube_microcerts():
...
+        attempts = []
+        edx_ids = next(
+            (
+                ids
+                for ids in product_list["externalIDs"]
+                if ids["origin"] == "EdX"
+            ),
+            None,
+        )
+        if edx_ids:
```
Comment (jpmartinspt) [inline webapp/cube/views.py:223]: Slightly confused by this, is there any listing that has more than one EDX id?

Comment (carkod) [reply webapp/cube/views.py:223]: Well that's why I wanted you to take a look at this PR. I get a list of IDs, although currently I just get one in the list. Should I just grab the first always `edx_ids[0]`?

Comment (jpmartinspt) [reply webapp/cube/views.py:223]: I believe so, if a product listing represents a single EDX course it should only have one EDX ID

Comment (jpmartinspt) [reply webapp/cube/views.py:223]: `edx_id = [edx_id for edx_id in product_list["externalIDs"] if ids["origin"] == "EdX"][0]`

Comment (carkod) [reply webapp/cube/views.py:223]: I asked CS team what do they do with the IDs and they said that they just synch Edx with UA contracts using the email. So I trust that they always will give at least one ID?

File: webapp/cube/views.py (hunk 3 above)
Comment (carkod) [reply webapp/cube/views.py:57]: Can I do instead `talisker.requests.get_session()`

Comment (jpmartinspt) [reply webapp/cube/views.py]: No, that will just end up reusing and existing session. You need to configure the created session.

File: webapp/app.py (hunk 1 above)
Comment (carkod) [reply webapp/app.py]: Done

File: webapp/cube/views.py (hunk 2 above)
Comment (carkod) [reply webapp/cube/views.py]: Done

File: webapp/cube/views.py (hunk 4 above)
Comment (carkod) [reply webapp/cube/views.py]: Good point. Done

File: webapp/cube/views.py (hunk 3 above)
Comment (carkod) [reply webapp/cube/views.py:57]: Done

**Comment by carkod** [comment]
Removed React code at Clem's request

File: webapp/cube/views.py [hunk 6]
```diff
@@ -133,6 +145,135 @@ def cube_microcerts():
...
+        course = {
+            "product_listing_id": product_list["id"],
+            "value": product_list["price"]["value"],
+        }
+        edx_ids = [
+            edx_id
+            for edx_id in product_list["externalIDs"]
+            if edx_id["origin"] == "EdX"
+        ][0]
+        for course_id in edx_ids["IDs"]:
```
Comment (jpmartinspt) [inline webapp/cube/views.py:213]: I think we can assume that edx_ids["IDs"] only has one value, so I would just do:
```
course_id = [
    edx_id[0]
    for edx_id in product_list["externalIDs"]
    if edx_id["origin"] == "EdX"
][0]
```
And then you can get rid of the for loop.

Comment (carkod) [reply webapp/cube/views.py]: mm ok. Let's hope they don't change the structure. There is another list inside, so I changed a bit the code.

**Review by ClementChaumel — APPROVED** [review]
LGTM

**Pull request merged**